File: src/2026/241. Different Ways to Add Parentheses/solution.py
class Solution:
    operations = {
        "+": lambda x,y : x+y,
        "-": lambda x,y : x-y,
        "*": lambda x,y : x*y,
    }

    def diffWaysToCompute(self, expression: str) -> List[int]:
        ans = []
        for index in range(len(expression)):
            if ( expression[index] in self.operations ):
                nums1 = self.diffWaysToCompute( expression[0:index])
                nums2 = self.diffWaysToCompute( expression[index+1:])
                for n1 in nums1:
                    for n2 in nums2:
                        ans.append(  self.operations[ expression[index] ](n1,n2) )
        if ans == []:
            return [ int(expression) ]
        return ans

File: src/2026/241. Different Ways to Add Parentheses/solution.cpp
class Solution {
public:
    inline int compute(char &op, int &x, int &y) {
        if (op == '+') {
            return x + y;
        } else if (op == '*') {
            return x * y;
        } else {
            return x - y;
        }
    }

    vector<int> diffWaysToCompute(string expression) {
        vector<int> ans;
        for (int i = 1; i < expression.length(); i++) {
            if (expression[i] == '+' || expression[i] == '-' ||
                expression[i] == '*') {
                auto left_hand_expr = expression.substr(
                    0, distance(expression.begin(), expression.begin() + i));
                auto left = diffWaysToCompute(left_hand_expr);
                auto right_hand_expr = expression.substr(
                    i + 1,
                    distance(expression.begin() + i + 1, expression.end()));
                auto right = diffWaysToCompute(right_hand_expr);

                for (auto& n1 : left) {
                    for (auto& n2 : right) {
                        ans.push_back(compute(expression[i], n1, n2));
                    }
                }
            }
        }
        if (ans.empty()) {
            return {stoi(expression)};
        }
        return ans;
    }
};
